src/compiler/diagnosticMessages.json

```
@@ -6227,6 +6227,11 @@
         "category": "Message",
         "code": 6718
     },
+    "Ensure that each file can have declaration emit generated without type information": {

```

```suggestion
    "Report errors on emitted declarations which rely excessively on type inference to produce the emit.": {

```
since
1. It doesn't actually "ensure" anything (since it's not adjusting the emit), it just adds errors (I'd have the same remark on the `isolatedModules` description - though that does sometimes affect emit, so it's less egregious). And
2. Type information (or something close enough to it - it's not like types are magical) is still being used, just only a subset of more local, simpler inferences (since it's not like we're issuing a diagnostic on *every* missing annotation, we do have a bunch of carve-outs that are "simple enough").

Does that reek of more ambiguity than the original message? Yes. But what this does is ambiguous - you really do need to know if it's reporting on what you want or not. For example, I doubt this is going to hit up everything you'd need to issue an error on for a JSR registry declaration emitter as it exists right now - the description they give is much more constrained than what we have here. This flag is basically a "set this if someone in your toolchain told you to" flag. Moreover, by keeping the description a bit more wishy-washy, when someone comes to us and says "Hey, `export const a = [1]`  is missing an annotation, why doesn't it have an error in `isolatedDeclarations`?" we can say "Well, the type's self-evident from the expression and we think any tool worth its' salt can synthesize a type node for it." without as much "oh, well, there's not _no_ inference going on because that'd be unusable in a laundry list of common examples" explaination. Plus, the ambiguity gives room for future changes, since we already *know* there's probably gonna be some.

---

That's probably a right enough description, since the upper limit of the desired behavior is deferring all imported names as opaque types/namespaces and deferring all (access, arithmetic, construction, etc) operations on them). Minor nit, ~source~ - just `files`. No need for the jargon that means the same thing. Maybe,
```suggestion
    "Report errors for missing type annotations whose creation requires nonlocal information.": {

```
since `import {N} from "a"; export const a = null as N;` is okay - it's not referring to other files that's conceptually banned by the flag, but leaking data from other files into this one that's banned.

---

Yeah, I was going to say "nonlocal" is ambiguous - someone could easily interpret "local" as meaning "not from an external dependency" when in fact it's more constrained than that.  The scope should be defined explicitly.

---

Changed to `Require sufficient annotation on exports so other tools can trivially generate declaration files.` May still change later